### ISSUE TITLE
Feature Request: Allow custom `setSelection` behavior at `doc` level

### ISSUE THREAD

**Issue by rev-hikari** [ProseMirror/prosemirror#878]
Current behavior of `setSelection` for block-level atomic node (`node.isBlock && node.isAtom`) is selecting the full dom node, which is insufficient in some case (e.g. use custom `nodeViews` to render complex dom tree inside it / try to hide weird cursor on iOS devices).

Since `nodeViews` is not suitable for `doc` node (as described in https://github.com/ProseMirror/prosemirror-view/pull/40#issuecomment-439343756), we may need another option to allow developers to customize the behavior for `setSelection` at `doc` level.

**Comment by marijnh** [comment]
Could you be more specific about the use cases? Why would a complex dom tree inside a node require you to override the way a selection around it is created? What kind of iOS cursor are you hiding?

**Comment by rev-hikari** [comment]
**1. Selecting a block will cause cursor leaves at the start of the line next to this block.**

Example: https://codesandbox.io/s/jprm772y25

![screen record](https://user-images.githubusercontent.com/44946699/50166916-ebb83480-0322-11e9-9e6c-1bb133b15be5.gif)

without styles for `ProseMirror-hideselection`:

![screen record](https://user-images.githubusercontent.com/44946699/50166854-d0e5c000-0322-11e9-9f67-32c9a639683e.gif)

**2. Selecting an image will show weird cursor on iOS**

Example: https://codesandbox.io/s/34p103vkz5

![screenshot](https://user-images.githubusercontent.com/44946699/50167512-4900b580-0324-11e9-8785-b9aeb7718901.png)

**Comment by marijnh** [comment]
>  Selecting a block will cause cursor leaves at the start of the line next to this block

What do you mean by "cursor leaves"? I still don't really understand what you are describing.

The iOS thing sounds like it should be handled by the core somehow, since that issue will come up for all users. What does your workaround look like, precisely?

**Comment by rev-hikari** [comment]
> What do you mean by "cursor leaves"? I still don't really understand what you are describing.

Sorry for my bad English.

![screen record](https://user-images.githubusercontent.com/44946699/50174200-853c1200-0334-11e9-8f32-a0c5a63070bd.gif)

Reproduce steps of the first bug:

1. Open https://codesandbox.io/s/jprm772y25
2. Click the selectable block
3. Click the paragraph next to the block

After I selected that block, the cursor is placed at the begin of the paragraph next to that block. At this moment, I cannot unselect that block by clicking the paragraph below, because it doesn't make any `selectionchange` event.

The main reason is, in this case, make a range to select the full block node is not permitted, and the browser will fallback to place the cursor next to that block.

To workaround these two bugs (this one and the iOS one above), we need change the dom structure to:

```html
<div class="wrapper" contenteditable="false">
  <div class="hidden-cursor"><!-- invisible char "\uFEFF" (this is important) --></div>
  <div class="content" draggable="true">
    <!-- real content goes here -->
  </div>
</div>
```

Select the block (on any platform except iOS):

```js
const selectBlock = node => {
  const selection = document.getSelection();
  const range = document.createRange();
  range.selectNode(node.querySelector(".content"));
  selection.removeAllRanges();
  selection.addRange(range);
}
```

Select the block on iOS:

```js
const selectBlock = node => {
  const selection = document.getSelection();
  const range = document.createRange();
  const hiddenCursor = node.querySelector(".hidden-cursor")
  range.setStart(hiddenCursor, 0);
  range.setEnd(hiddenCursor, 0);
  selection.removeAllRanges();
  selection.addRange(range);
}
```

Styles:

```css
.wrapper {
  position: relative;
}

.hidden-cursor {
  position: absolute;
  left: 0;
  top: 0;
  font-size: 0;
}
```

**Comment by rev-hikari** [comment]
Monkey patch the `view.docView.setSelection` works fine with the workaround described above.

**Comment by rev-hikari** [comment]
Anyway, allow developers to customize the behavior for `setSelection` is not meaningless, even after these bugs were fixed in the core module.

**Comment by marijnh** [comment]
> The main reason is, in this case, make a range to select the full block node is not permitted, and the browser will fallback to place the cursor next to that block.

Which browser is this? I thought we had a functioning workaround for Chrome now, and other browsers allowed such selections the last time I tested.

Workaround like the one you show are highly invasive (you'll trigger DOM mutation events when making the DOM changes, for example, and/or confuse the library when it is parsing content from the DOM), and creating an interface that allows client code to do this without being extremely error-prone is probably not that easy.

**Comment by rev-hikari** [comment]
I tested it in my chrome browser. Is the example above working fine on your
computer?

**Comment by rev-hikari** [comment]
I know it's invasive, implementing it in the core module may not be the
best way. At least, we should allow developers to customize the behavior of
`setSelection`, so they can do something to resolve it using their own ways
without monkey-patching anything.

**Comment by marijnh** [comment]
> Is the example above working fine on your computer?

Works in Firefox, but Chrome indeed seems to have regressed. I'll take a look.

> I know it's invasive, implementing it in the core module may not be the best way. At least, we should allow developers to customize the behavior of `setSelection`, so they can do something to resolve it using their own ways without monkey-patching anything.

It being invasive can be a good argument for implementing in core—just providing a hook that invites people to do broken stuff isn't a good idea, and since this kind of hacks seem to be the only use case for this feature, I'd rather not provide it.

**Comment by marijnh** [comment]
So the selection as returned by `getSelection` in this case spans from index 1 to 2 in the top-level editable node. But for some reason Chrome _shows_ it as a cursor in the paragraph after the image, and thus doesn't do anything when you click near where that cursor is. I guess it is another case of the selection reported to the script not agreeing with the actual selection, which I've seen before in Chrome.

Attached patch adds a kludge to the mouse event handling to notice this and work around it. Does that resolve that part of the problem for you?

**Comment by marijnh** [comment]
As for the other part, I haven't found a workaround yet—I guess the problem happens because Safari draws a handle for the selection, even though it's not really a text selection. By creating a custom element that's absolutely positioned, and moving the selection there, you convince it not to do this.

I guess you already tried setting `-webkit-touch-callout: none` on the element?

**Comment by rev-hikari** [comment]
> I guess you already tried setting `-webkit-touch-callout: none` on the element?

I have tried it, and it doesn't have any effect.

> Attached patch adds a kludge to the mouse event handling to notice this and work around it. Does that resolve that part of the problem for you?

It may resolve that problem in chrome, but in combination of these two bugs:

![screenshot from 2018-12-18 23-20-31](https://user-images.githubusercontent.com/44946699/50213689-4c924c00-03b9-11e9-8e04-204727f07f06.png)

Weird cursor still shown on iOS devices.

**Comment by rev-hikari** [comment]
The first problem not only happened in Chrome, but it also happened in Safari (both iOS and macOS).

**Comment by marijnh** [comment]
Interesting. Desktop Safari didn't seem to show this bug, but I guess it either shows it slightly differently or it only happens in Mobile Safari.

I'm not going to have time to work on this further on the short term, and I stand by my position that just adding a hook here isn't a good idea, so for now I recommend you continue monkey-patching.

(It might also be worthwhile to try some variations in the node's HTML structure or styling—I think you're the first to run into this issue, since most uneditable block nodes do not trigger the browser behavior.)

**Comment by rev-hikari** [comment]
https://nightingale-editor.netlify.com/

Here's the POC of that solution described above.

**Comment by rev-hikari** [comment]
Source code: https://github.com/maltose-lab/nightingale/blob/master/src/plugins/react/createReactPlugin.js#L52-L80

**Comment by rev-hikari** [comment]
| Before | After |
| - | - |
| ![](https://user-images.githubusercontent.com/44946699/56181050-734a5180-603e-11e9-8a44-0a3e2d1ba549.png) | ![](https://user-images.githubusercontent.com/44946699/56181049-734a5180-603e-11e9-98ab-2fbe7cdd87a1.png) |

**Comment by matolg** [comment]
Hi @marijnh.
I faced the similar issue (with carret appearing under the selected block) on iOS Safari and react-native-webview component.
I've noticed it in my app and then was able to repro this with https://codesandbox.io/s/jprm772y25 example just opening it in Safari on iOS 13 (or 12).

Here is **iOS Safari** demo:
![Actual - Safari iOS - react-native-webview](https://user-images.githubusercontent.com/1191920/71074595-9f9d0b80-2193-11ea-9a44-bf5f060a3a3a.gif)
This is for iOS Safari only.

On **Desktop Safari** there is no caret under the block, but it is impossible to put caret to the next line (paragraph) once block was selected though:
![Actual - Safari macOS](https://user-images.githubusercontent.com/1191920/71074819-028ea280-2194-11ea-8be8-275f03013ea9.gif)

On **Chrome** it works as expected:
![Expected - Chrome](https://user-images.githubusercontent.com/1191920/71074882-23ef8e80-2194-11ea-871d-92353051a37c.gif)
